node/util_test.go

```diff
@@ -0,0 +1,123 @@
+package node
+
+import (
+	"context"
+	"errors"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/config"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+// NewTestNetwork creates a network of fully connected nodes.
+func NewTestNetwork(t *testing.T, conf config.Config, l log.Log, size int) []*TestApp {
+	// We need to set this global state
+	types.SetLayersPerEpoch(conf.LayersPerEpoch)
+	types.SetNetworkHRP(conf.NetworkHRP) // set to generate coinbase
+
+	// To save an epoch of startup time, we bootstrap (meaning we manually set
+	// it) the beacon for epoch 2 so that in epoch 3 hare can start.
+	bootstrapEpoch := (types.GetEffectiveGenesis() + 1).GetEpoch()
+	bootstrapBeacon := types.Beacon{}
+	genesis := conf.Genesis.GenesisID()
+	copy(bootstrapBeacon[:], genesis[:])
+
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	g := errgroup.Group{}
```

Comment:
This limits any test to at most 20 seconds before all nodes are forcefully shut down. I think just a `cancelCtx` would fit better:

```suggestion
	ctx, cancel := context.WithCancel(context.Background())
	var g errgroup.Group
```

Reply:
It would give the caller more flexibility, but also increases the burden on the caller.

Given that this is a test component I can't imagine a situation where you want to cancel the context for the network start because you already got what you needed through some other route. So I think it would boil down to callers providing a context with a timeout that they choose based on the size of the network they are creating.

On my machine the time taken for the network to start up is not that long, about 1.2 s  for a network of 3, 5s for a network of 10 and 16s for a network of 20. So in normal use I don't think that deadline will ever be reached, since this probably isn't a good fit for networks of size greater than 5 or 6, 20 caused my browsers to crash :) The deadline is there just to avoid waiting for 5 minutes if for some reason the network startup hangs.

Reply:
The context isn't used correctly by `app.Start`. It determines not the timeout to start the node, but rather its whole lifetime. If `ctx` is canceled, all nodes shut down.

This means the way `NewTestNetwork` is implemented at the moment doesn't allow any tests that run longer than 20 seconds (including startup and the test itself), because by then all nodes will forcefully be shut down (by canceling the context), independent of if the test finished or not.

Changing it in the way I suggested it gives the nodes infinite lifetime (and they will still be shut down when the test ends by the `cancel()` call in the `t.Cleanup` function.

For the caller it makes no difference, besides that a test can be longer than 20 seconds 🙂

Reply:
Hey, I misunderstood your original suggestion. Thank you for clarifying. I've removed the timeout :+1: